policy/aggregation_type.go

```diff
@@ -0,0 +1,238 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package policy
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/m3db/m3metrics/generated/proto/schema"
+)
+
+// Supported aggregation types.
+const (
+	// Not using iota for better readability.
+	Unknown AggregationType = 0
+	Last    AggregationType = 1
+	Lower   AggregationType = 2
+	Upper   AggregationType = 3
+	Mean    AggregationType = 4
+	Median  AggregationType = 5
+	Count   AggregationType = 6
+	Sum     AggregationType = 7
+	SumSq   AggregationType = 8
+	Stdev   AggregationType = 9
+	P50     AggregationType = 10
+	P95     AggregationType = 11
+	P99     AggregationType = 12
+	P999    AggregationType = 13
+	P9999   AggregationType = 14
+)
+
+const (
+	// TotalAggregationTypes is the number of supported aggregation types.
+	// NB(cw) this should be updated if new AggregationTypes were added.
+	TotalAggregationTypes = 14
+
+	// CompressedSize is the length of the CompressedAggregationTypes.
+	// The CompressedSize will be 1 when totalAggregationTypes <= 63.
+	CompressedSize = TotalAggregationTypes/64 + 1
+
+	aggregationTypesSeparator = ","
+)
+
+var (
+	// EmptyAggregationTypes is an empty list of AggregationTypes.
+	EmptyAggregationTypes AggregationTypes
+
+	// EmptyCompressedAggregationTypes is an empty CompressedAggregationTypes.
+	EmptyCompressedAggregationTypes CompressedAggregationTypes
+
+	aggregationTypeStringMap map[string]AggregationType
+)
+
+func init() {
+	aggregationTypeStringMap = make(map[string]AggregationType, TotalAggregationTypes)
+	for i := 1; i <= TotalAggregationTypes; i++ {
+		aggType := AggregationType(i)
+		aggregationTypeStringMap[aggType.String()] = aggType
+	}
+}
+
+// AggregationType defines a custom aggregation function.
+type AggregationType int
+
+// NewAggregationTypeFromSchema creates an aggregation type from a schema.
+func NewAggregationTypeFromSchema(input schema.AggregationType) (AggregationType, error) {
+	aggType := AggregationType(input)
+	if !aggType.IsValid() {
+		return Unknown, fmt.Errorf("invalid aggregation type from schema: %s", input)
+	}
+	return aggType, nil
+}
+
+// IsValid checks if an AggregationType is valid.
+func (a AggregationType) IsValid() bool {
+	return a > 0 && a <= TotalAggregationTypes
+}
+
+// IsValidForGauge if an AggregationType is valid for Gauge.
+func (a AggregationType) IsValidForGauge() bool {
+	switch a {
+	case Last, Lower, Upper, Mean, Count, Sum, SumSq, Stdev:
+		return true
+	default:
+		return false
+	}
+}
+
+// IsValidForCounter if an AggregationType is valid for Counter.
+func (a AggregationType) IsValidForCounter() bool {
+	switch a {
+	case Lower, Upper, Mean, Count, Sum, SumSq, Stdev:
+		return true
+	default:
+		return false
+	}
+}
+
+// IsValidForTimer if an AggregationType is valid for Timer.
+func (a AggregationType) IsValidForTimer() bool {
+	switch a {
+	case Last:
```

Comment:
Okay, this would require special logic on the UI, although I guess that's needed for not allowing percentiles for counters/gauges anyway.